clang/lib/Sema/SemaDecl.cpp

```diff
@@ -2282,7 +2282,8 @@ void Sema::ActOnPopScope(SourceLocation Loc, Scope *S) {
 
     // Remove this name from our lexical scope, and warn on it if we haven't
     // already.
-    IdResolver.RemoveDecl(D);
+    if (!PP.isIncrementalProcessingEnabled())
+      IdResolver.RemoveDecl(D);
```

Comment:
so in incremental processing mode, we never remove declarations from the id resolver, so they always stay available? e.g.,
```
{
  int i = 12;
}
printf("%d", i); // Fine?
```

Reply:
No, this change makes sure that we preserve the state of the `IdResolver` across partial translation units (PTU). That is, when we see another incremental output, we still take into account the declarations that came from previous PTUs. Eg:

```
[repl] int i = 12;
[repl] printf("%d\n", i); // without this patch we report we do not know what `i` is...
```

To your explicit question this patch produces:

```
bin/clang-repl -Xcc -xc
clang-repl> { int i = 12; }
clang-repl> printf("%d", i); // Fine?
In file included from <<< inputs >>>:1:
input_line_2:1:1: error: call to undeclared library function 'printf' with type 'int (const char *, ...)'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
    1 | printf("%d", i); // Fine?
      | ^
input_line_2:1:1: note: include the header <stdio.h> or explicitly provide a declaration for 'printf'
input_line_2:1:14: error: reference to local variable 'i' declared in enclosing context
    1 | printf("%d", i); // Fine?
      |              ^
input_line_1:1:7: note: 'i' declared here
    1 | { int i = 12; }
      |       ^
error: Parsing failed.
clang-repl> %quit
```

Reply:
Ah thank you for the explanation -- I think that warrants a comment in the code to explain what's going on, because it's not obvious from context.

Reply:
Added a comment.